File: 22-generate-parentheses/generate-parentheses.cpp
class Solution {
public:
    vector<string> generateParenthesis(int n) {
        
        vector<string> res;

        dfs(0, 0, n, res, "");

        return res;
    }

    void dfs(int l, int r, int t, vector<string>& res, string s) {
        if(l==r && (l+r) == t*2) {
            res.push_back(s);
            return;
        }

        if(l<t) {
            dfs(l+1, r, t, res, s + '(');
        }

        if(r<l) {
            dfs(l, r+1, t, res, s + ')');
        }

    }
};

File: 22-generate-parentheses/generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        def backtrack(open_count, close_count, curr, ans):
            if open_count == n and close_count == n:
                ans.append(curr)
                return

            if open_count < n:
                backtrack(open_count + 1, close_count, curr + "(", ans)

            if close_count < open_count:
                backtrack(open_count, close_count + 1, curr + ")", ans)

        ans = []
        backtrack(0, 0, "", ans)
        return ans